### ISSUE TITLE
Deprecate using (internal) convention mapping for properties of type `Provider`

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle#8404]
`Provider`, `Property` and other lazy types such as `FileCollection` are intended to represent some calculated value. It doesn't make sense for convention mappings to be applied to properties of these types, as the calculation should instead be carried by the lazy type itself.

### Expected Behavior

Gradle fails eagerly when a convention mapping is applied to a property whose type is some lazy type.

### Current Behavior

Calculated values can be implemented using convention mappings, but doing so has a bunch of downsides. For example, the value may be calculated many times, which is a performance issue and may lead to different consumers seeing different values at different times.

See also #8401.

**Comment by lacasseio** [comment]
FYI: Mismatch of return type from the mapping closure ignores the mapping. It's internal so we may not care but I spend some time figuring that out.

**Comment by lacasseio** [comment]
It seems convention mapping on `Property` is completely ignored, am I seeing the right behaviour? //cc @adammurdoch 

If yes, I'm thinking of just throw an exception for all `Property` type (`ListProperty`, `SetProperty`, `MapProperty`, `Property`) with an error message of just using `Property#set` to register a `Provider`.

**Comment by big-guy** [comment]
@lacasseio do you have any examples/test you've setup to show this?  I know I've seen it being used with `Provider`, but I'm not sure about `Property`.

I think the best place for this check to go is when someone registers the mapping.

**Comment by lacasseio** [comment]
I agree. I have some teststhat I will push soon. I didn't believe it at first but like you said, Provider mapping works but it doesn't work with Property.

**Comment by lacasseio** [comment]
PR https://github.com/gradle/gradle/pull/8580 throw an exception for convention mapping of types like `Property`, `ListProperty`, `SetProperty` and `MapProperty`. The convention mapping wasn't working and was silently ignored.

**Comment by lacasseio** [comment]
I propose printing a deprecation warning for `Provider` and `FileCollection`/`FileTree` to mention those are read-only. They should use setter if available or open issue to convert them to `Property` or configurable things.

I propose printing a deprecation warning for `ConfigurableFileCollection` and `ConfigurableFileTree`  to mention to use `from` or `setFrom` instead.

What do you think?

**Comment by adammurdoch** [comment]
I'd be strict rather than lenient for `Provider` typed properties, if we can. For `FileCollection` typed properties, we really need to be lenient for backwards compatibility. We should deal with `FileCollection` properties as a separate issue, I think, and first look at our own violations of the rules and figure out the replacements before we start complaining about other people's usages.

In general, it doesn't really make sense to disallow a property to be mutated using a convention mapping but still allow it to be mutated using a setter. So, I think we should deprecate/fail both together.

We should also take `@ReplacedBy` into account in the validation. For example, we might not complain about a setter on a property that is replaced by another, but complain about it on a property that is not replaced.

**Comment by lacasseio** [comment]
Good point, let me see what I can do, around this.

**Comment by lacasseio** [comment]
All my changes are parked in [this branch](https://github.com/gradle/gradle/tree/lacasseio/deprecate-convention-mapping-for-provider).

**Comment by big-guy** [comment]
We should check that this doesn't break nebula's os-package plugin.  It extends our AbstractArchiveTask types and tries to override Provider/Property with convention mapping.
